synapse/storage/databases/main/registration.py

```diff
@@ -1920,6 +1920,27 @@ async def consume_login_token(self, token: str) -> LoginTokenLookupResult:
             self._clock.time_msec(),
         )
 
+    async def invalidate_login_tokens_by_session_id(
+        self, auth_provider_id: str, auth_provider_session_id: str
+    ) -> None:
+        """Invalidate login tokens with the given IdP session ID.
+
+        Args:
+            auth_provider_id: The SSO Identity Provider that the user authenticated with
+                to get this token
+            auth_provider_session_id: The session ID advertised by the SSO Identity
+                Provider
+        """
+        await self.db_pool.simple_update(
+            table="login_tokens",
+            keyvalues={
+                "auth_provider_id": auth_provider_id,
+                "auth_provider_session_id": auth_provider_session_id,
+            },
+            updatevalues={"used_ts": self._clock.time_msec()},
```

Comment:
Setting it as used invalidates it?

(Naively I was expecting a delete here.)

Reply:
Yeah, it does. Initially in #13844 I was deleting them, but I moved to a `used_ts` field to be able to track token reuse and use after expiration in a prometheus counter. Maybe that could be a separate column though, like `revoked_at`, because it means that logins cancelled in-flight will be considered as "token reuse" in that counter 🤷 
We can always change that later

Reply:
Thanks for the explanation!